README.md
# d-ploy README

This is the README for your extension "d-ploy". After writing up a brief description, we recommend including the following sections.

## Features

Describe specific features of your extension including screenshots of your extension in action. Image paths are relative to this README file.

For example if there is an image subfolder under your extension project workspace:

\!\[feature X\]\(images/feature-x.png\)

> Tip: Many popular extensions utilize animations. This is an excellent way to show off your extension! We recommend short, focused animations that are easy to follow.

## Requirements

If you have any requirements or dependencies, add a section describing those and how to install and configure them.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

* `myExtension.enable`: enable/disable this extension
* `myExtension.thing`: set to `blah` to do something

## Known Issues

Calling out known issues can help limit users opening duplicate issues against your extension.

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z.

-----------------------------------------------------------------------------------------------------------
## Following extension guidelines

Ensure that you've read through the extensions guidelines and follow the best practices for creating your extension.

* [Extension Guidelines](https://code.visualstudio.com/api/references/extension-guidelines)

## Working with Markdown

**Note:** You can author your README using Visual Studio Code.  Here are some useful editor keyboard shortcuts:

* Split the editor (`Cmd+\` on macOS or `Ctrl+\` on Windows and Linux)
* Toggle preview (`Shift+CMD+V` on macOS or `Shift+Ctrl+V` on Windows and Linux)
* Press `Ctrl+Space` (Windows, Linux) or `Cmd+Space` (macOS) to see a list of Markdown snippets

### For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "d-ploy" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import DployPanel from './DployPanel/DployPanel';

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {

	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	console.log('Congratulations, your extension "d-ploy" is now active!');

	// The command has been defined in the package.json file
	// Now provide the implementation of the command with registerCommand
	// The commandId parameter must match the command field in package.json
	let disposable = vscode.commands.registerCommand('d-ploy.launch', () => {
		let dployPanel = new DployPanel(context)
		dployPanel.init()
		dployPanel.setWebViewHTML()
	});

	context.subscriptions.push(disposable);
}

// this method is called when your extension is deactivated
export function deactivate() {}


src/Panel/Panel.ts
import * as vscode from 'vscode'
// import iPanel from './iPanel'

export default abstract class Panel {
    protected context: vscode.ExtensionContext
    protected panel: vscode.WebviewPanel | undefined
    protected readonly viewType: string
    protected readonly title: string
    protected readonly showOptions: vscode.ViewColumn | {preserveFocus: boolean, viewColumn: vscode.ViewColumn}
    protected options?: vscode.WebviewPanelOptions & vscode.WebviewOptions = {}

    constructor(
      context: vscode.ExtensionContext,
      viewType: string,
      title: string,
      showOptions: vscode.ViewColumn | {preserveFocus: boolean, viewColumn: vscode.ViewColumn},
      options?: vscode.WebviewPanelOptions & vscode.WebviewOptions) {
      this.context = context
      this.viewType = viewType;
      this.title = title;
      this.showOptions = showOptions;
    }

    init(): void {
      this.panel = vscode.window.createWebviewPanel(this.viewType, this.title, this.showOptions, this.options);
      console.log('initiated')
    }

    abstract get webViewHTML() : string

    setWebViewHTML() : void {
      if (this.panel) {
        this.panel.webview.html = this.webViewHTML
        console.log('html loaded')
      }
    }
}

src/DployPanel/dployPanelHTML.ts
import * as vscode from 'vscode'

const onSubmit = (e: object) => {
    console.log(e)
    vscode.window.showInformationMessage('Submitted!')
}

const dployPanelHTML = () => `
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>D-ploy</title>
    <style>
        label, input[type=text], select, textarea {
            width: 100%;
            color: #f5f5f5;
            outline: none;
            display: block;
        }

        input, select, textarea, button {
            background-color: #333;
            color: #f5f5f5;
            border: 2px solid #555;
            margin-bottom: 1rem;
            padding: .4rem;
        }

        textarea {
            height: 250px;
            resize: none;
        }

        input[type=radio] + label,
        label.radio {
            display: inline
        }

        .container {
            width: 625px;
            margin: auto;
        }
    </style>
</head>
<body>
    <h1>Create your D-ployment</h1>
    <div class="container">
        <form onsubmit="onSubmit">
            <label for="deployTo">Branch to deploy to</label>
            <input type="text" id="deployTo" name="deployTo" />

            <!--<label for="mergeType">Select merge type</label>-->
            <select type="text" id="mergeType" name="mergeType">
                <option>Select merge type</option>
                <option>Merge</option>
                <option>Fast-forward</option>
                <option>No Fast-forward</option>
                <option>Squash</option>
            </select>

            <label for="commitMessage">Commit message</label>
            <textarea type="text" id="commitMessage" name="commitMessage"></textarea>

            <label>Create tag?</label>
            <input type="radio" id="tagYes" name="createTag" value="true" />
            <label for="tagYes" class="radio">Yes</label>
            <input type="radio" id="tagNo" name="createTag" value="false" />
            <label for="tagNo" class="radio">No</label>

            <label for="tagName">Tag name</label>
            <input type="text" id="tagName" name="tagName" />

            <label for="tagMessage">Tag Message</label>
            <textarea type="text" id="tagMessage" name="tagMessage"></textarea>        

            <input type="submit" value="D-ploy!" />

            <script type="text/javascript">
                const onSubmit = (e: object) => {
                    console.log(e)
                    vscode.window.showInformationMessage('Submitted!')
                }
            </script>
        </form>
    </div>
</body>
</html>
`

export default dployPanelHTML

src/DployPanel/DployPanel.ts
import * as vscode from 'vscode'
import Panel from "../Panel/Panel";
import dployPanelHTML from './dployPanelHTML';

export default class DployPanel extends Panel {
    constructor(context: vscode.ExtensionContext) {
        super(
            context,
            'd-ploy', 
            'D-ploy', 
            vscode.ViewColumn.One, 
            // dployPanelHTML(),
            {enableScripts: true, retainContextWhenHidden: true}
        )
    }

    get webViewHTML() {
        return `
          <!DOCTYPE html>
          <html lang="en">
          <head>
              <meta charset="UTF-8">
              <meta name="viewport" content="width=device-width, initial-scale=1.0">
              <title>D-ploy</title>
              <style>
                  label, input[type=text], select, textarea {
                      width: 100%;
                      color: #f5f5f5;
                      outline: none;
                      display: block;
                  }
          
                  input, select, textarea, button {
                      background-color: #333;
                      color: #f5f5f5;
                      border: 2px solid #555;
                      margin-bottom: 1rem;
                      padding: .4rem;
                  }
          
                  textarea {
                      height: 250px;
                      resize: none;
                  }
          
                  input[type=radio] + label,
                  label.radio {
                      display: inline
                  }
          
                  .container {
                      width: 625px;
                      margin: auto;
                  }
              </style>
          </head>
          <body>
              <h1>Create your D-ployment</h1>
              <div class="container">
                  <form onsubmit="onSubmit">
                      <label for="deployTo">Branch to deploy to</label>
                      <input type="text" id="deployTo" name="deployTo" />
          
                      <!--<label for="mergeType">Select merge type</label>-->
                      <select type="text" id="mergeType" name="mergeType">
                          <option>Select merge type</option>
                          <option>Merge</option>
                          <option>Fast-forward</option>
                          <option>No Fast-forward</option>
                          <option>Squash</option>
                      </select>
          
                      <label for="commitMessage">Commit message</label>
                      <textarea type="text" id="commitMessage" name="commitMessage"></textarea>
          
                      <label>Create tag?</label>
                      <input type="radio" id="tagYes" name="createTag" value="true" />
                      <label for="tagYes" class="radio">Yes</label>
                      <input type="radio" id="tagNo" name="createTag" value="false" />
                      <label for="tagNo" class="radio">No</label>
          
                      <label for="tagName">Tag name</label>
                      <input type="text" id="tagName" name="tagName" />
          
                      <label for="tagMessage">Tag Message</label>
                      <textarea type="text" id="tagMessage" name="tagMessage"></textarea>        
          
                      <input type="submit" value="D-ploy!" />
          
                      <script type="text/javascript">
                          const onSubmit = (e: object) => {
                              console.log(e)
                              vscode.window.showInformationMessage('Submitted!')
                          }
                      </script>
                  </form>
              </div>
          </body>
          </html>
        `
      }
}